components/raft_log_engine/src/engine.rs

```diff
@@ -25,14 +29,143 @@ impl MessageExt for MessageExtTyped {
     }
 }
 
+struct ManagedReader<R: Seek + Read> {
+    raw: Option<R>,
+    decrypter: Option<DecrypterReader<R>>,
+    rate_limiter: Option<Arc<IORateLimiter>>,
+}
+
+impl<R: Seek + Read> Seek for ManagedReader<R> {
+    fn seek(&mut self, pos: SeekFrom) -> IoResult<u64> {
+        match (&mut self.raw, &mut self.decrypter) {
+            (Some(ref mut reader), None) => reader.seek(pos),
+            (None, Some(ref mut reader)) => reader.seek(pos),
+            _ => unreachable!(),
+        }
+    }
+}
+
+impl<R: Seek + Read> Read for ManagedReader<R> {
+    fn read(&mut self, buf: &mut [u8]) -> IoResult<usize> {
+        let mut size = buf.len();
+        if let Some(ref mut limiter) = self.rate_limiter {
+            size = limiter.request(IOType::ForegroundRead, IOOp::Read, size);
+        }
+        match (&mut self.raw, &mut self.decrypter) {
+            (Some(ref mut reader), None) => reader.read(&mut buf[..size]),
+            (None, Some(ref mut reader)) => reader.read(&mut buf[..size]),
+            _ => unreachable!(),
+        }
```

Comment:
how about
```
let read_size = match {
};
limiter.request(read_size);
```

Reply:
We put the `request` before actual IOs in other call sites too (std::File, rocksdb::Env), right now let's keep it consistent. But I agree the ideal placement is an open question, will look into it when we work on rate limiting later.